### PR TITLE
contributing: Explain how to run common tests

### PR DESCRIPTION
**Pull request by l0b0**
###### Description of changes

As a beginning contributor, it is far from obvious where to find tests,
how to run individual tests locally for quick development, and which
tests do what. The new "Testing changes" section aims to help with all
of these.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -61,6 +61,15 @@ Pull requests should not be squash merged in order to keep complete commit messa
 This means that, when addressing review comments in order to keep the pull request in an always mergeable status, you will sometimes need to rewrite your branch's history and then force-push it with `git push --force-with-lease`.
 Useful git commands that can help a lot with this are `git commit --patch --amend` and `git rebase --interactive`. For more details consult the git man pages or online resources like [git-rebase.io](https://git-rebase.io/) or [The Pro Git Book](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History).
 
+## Testing changes
+
+To run the main types of tests locally:
+
+- Run [NixOS tests](https://nixos.org/manual/nixos/unstable/#sec-nixos-tests) with `nix-build nixos/tests/NAME.nix`, where `NAME` is the name of a specific test file
```
Comment (drupol) [inline CONTRIBUTING.md:68]: Should we use `nix-build` or `nix build` ?

**Review by drupol — COMMENTED** [review]
Hi!

I definitely welcome this change, I just added a minor comment.

File: CONTRIBUTING.md (hunk 1 above)
Comment (l0b0) [reply CONTRIBUTING.md:68]: `nix-build`, because `nix build` is still experimental.

Comment (drupol) [reply CONTRIBUTING.md:68]: Fair enough.

**Review by emilylange — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -61,6 +61,15 @@ Pull requests should not be squash merged in order to keep complete commit messa
...
+- Run [global package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests) with `nix-build --attr PACKAGE pkgs/test/default.nix`, where `PACKAGE` is the package name
+- Run package-internal tests with `nix-build --attr pkgs.PACKAGE.passthru.tests`
+- Run library tests with `nix-build lib/tests/NAME.nix`
```
[the remark this answers is not in the document]
Comment (l0b0) [reply CONTRIBUTING.md:71]: `nix-build lib/tests/misc.nix` also passes. Is it possible that the comment is obsolete?

[the remark this answers is not in the document]
Comment (l0b0) [reply CONTRIBUTING.md]: Thank you! How's the current result?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -61,6 +61,15 @@ Pull requests should not be squash merged in order to keep complete commit messa
 This means that, when addressing review comments in order to keep the pull request in an always mergeable status, you will sometimes need to rewrite your branch's history and then force-push it with `git push --force-with-lease`.
 Useful git commands that can help a lot with this are `git commit --patch --amend` and `git rebase --interactive`. For more details consult the git man pages or online resources like [git-rebase.io](https://git-rebase.io/) or [The Pro Git Book](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History).
 
+## Testing changes
+
+To run the main types of tests locally:
+
+- Run [NixOS tests](https://nixos.org/manual/nixos/unstable/#sec-nixos-tests) with `nix-build nixos/tests/NAME.nix`, where `NAME` is the name of a specific test file
+- Run [global package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests) with `nix-build --attr PACKAGE pkgs/test/default.nix`, where `PACKAGE` is the package name
+- Run package-internal tests with `nix-build --attr pkgs.PACKAGE.passthru.tests`
```
[the remark this answers is not in the document]
Comment (drupol) [reply CONTRIBUTING.md:68]: That doesn't look like it is resolved.

[the remark this answers is not in the document]
Comment (l0b0) [reply CONTRIBUTING.md:68]: I resolved the "This should probably be at the top of the list, with some note that this is the most common thing." suggestion. `-A` and `--attr` have identical semantics, but the latter is clearer.

[the remark this answers is not in the document]
Comment (drupol) [reply CONTRIBUTING.md:68]: Fair enough. Setting back to resolved now.

**Review by drupol — APPROVED** [review]

**Pull request merged**